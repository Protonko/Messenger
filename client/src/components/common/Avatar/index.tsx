import {useMemo, FC} from 'react'
import classNames from 'classnames'
import {Sizes} from 'models/common/sizes'
import {ColorGeneratorByName} from 'utils/ColorGeneratorByName'
import {isDarkColor} from 'utils/isDarkColor'
import {Loader} from 'components/common/Loader'

export interface IPropsAvatar {
  src: string
  name?: string
  size?: Sizes
  customStyles?: string
}

export const Avatar: FC<IPropsAvatar> = ({
  src,
  name,
  size = Sizes.MEDIUM,
  customStyles,
}) => {
  const backgroundColor = useMemo(
    () => new ColorGeneratorByName(name ?? '').generate(),
    [src, name],
  )
  const avatarClassname = classNames(
    'avatar',
    {'avatar--sm': size === Sizes.SMALL},
    {'avatar--lg': size === Sizes.LARGE},
    {[customStyles ?? '']: customStyles},
  )

  if (!src && name) {
    const classNamesText = classNames(
      'avatar__text',
      {'avatar__text--light': isDarkColor(backgroundColor)},
      {'avatar__text--dark': !isDarkColor(backgroundColor)},
    )

    return (
      <div className={avatarClassname} style={{backgroundColor}}>
        <span className={classNamesText}>{name.charAt(0)}</span>
      </div>
    )
  }

  return (
    <div className={avatarClassname}>
      <img className="avatar__image" src={src} alt="" />
      <Loader />
    </div>
  )
}
