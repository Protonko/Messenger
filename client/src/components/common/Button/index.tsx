import {FC, HTMLProps, ReactNode} from 'react'

interface IButtonProps extends HTMLProps<HTMLButtonElement> {
  text?: string
  icon?: ReactNode
  leftAddons?: ReactNode
  rightAddons?: ReactNode
  type?: 'button' | 'submit' | 'reset'
}

export const Button: FC<IButtonProps> = ({
  text,
  icon,
  leftAddons,
  rightAddons,
  ...buttonProps
}) => {
  const renderButtonContent = () => {
    return (
      <>
        {text && (
          <span className="button__text">
            {text}
          </span>
        )}
        {icon && (
          <span className="button__icon">
            {icon}
          </span>
        )}
      </>
    )
  }

  return (
    <button className="button" {...buttonProps}>
      {leftAddons}
      {renderButtonContent()}
      {rightAddons}
    </button>
  )
}
