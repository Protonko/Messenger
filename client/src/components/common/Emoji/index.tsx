import 'emoji-mart/css/emoji-mart.css'
import {FC, useState, useRef} from 'react'
import {EmojiData, Picker} from 'emoji-mart'
import {ReactComponent as Smile} from 'assets/icons/smile.svg'
import {useOutsideClick} from 'hooks/useOutsideClick'

export interface IPropsEmoji {
  onSelect: (emoji: EmojiData) => void
}

export const Emoji: FC<IPropsEmoji> = ({onSelect}) => {
  const [open, setOpen] = useState(false)
  const ref = useRef(null)
  const initiatorRef = useRef(null)

  useOutsideClick(ref, () => setOpen(false), initiatorRef)

  const onSelectEmoji = (emoji: EmojiData) => {
    onSelect(emoji)
    setOpen(false)
  }

  return (
    <div className="emoji">
      {open && (
        <div className="emoji__popup" ref={ref}>
          <Picker native={true} onSelect={onSelectEmoji} />
        </div>
      )}

      <Smile
        className="emoji__button"
        ref={initiatorRef}
        onClick={() => setOpen(!open)}
      />
    </div>
  )
}
