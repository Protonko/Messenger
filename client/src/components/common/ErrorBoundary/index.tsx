import {Component, ErrorInfo} from 'react';

interface IProps {}
interface IState {
  error: null | Error;
  errorInfo: null | ErrorInfo;
  hasError: boolean;
}

export class ErrorBoundary extends Component<IProps, IState> {
  constructor(props: IProps) {
    super(props);

    this.state = {
      error: null,
      errorInfo: null,
      hasError: false,
    }
  }

  static getDerivedStateFromError(error: Error) {
    console.log(error)
    return {
      hasError: true,
    };
  }

  componentDidCatch(error: Error, errorInfo: ErrorInfo) {
    this.setState({
      error: error,
      errorInfo: errorInfo
    })
  }

  _renderDetailInfo() {
    if (process.env.NODE_ENV === 'development') {
      return (
        <div>

        </div>
      )
    }

    return null
  }

  render() {
    if (this.state.hasError) {
      return (
        <>
          <h1>Что-то пошло не так.</h1>

          {process.env.NODE_ENV === 'development' && (
            <details style={{ whiteSpace: 'pre-wrap' }}>
              {this.state.error && this.state.error.toString()}
              <br />
              {this.state.errorInfo?.componentStack}
            </details>
          )}
        </>
      )
    }

    return this.props.children;
  }
}
