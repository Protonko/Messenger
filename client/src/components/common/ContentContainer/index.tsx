import type {FC} from 'react'
import {Loader} from 'components/common/Loader'
import {Text} from 'components/common/Text'

interface ContentContainerProps {
  loading: boolean
  errorMessage: string | null
  children: JSX.Element
}

export const ContentContainer: FC<ContentContainerProps> = ({
  loading,
  errorMessage,
  children,
}) => {
  if (errorMessage) {
    return <Text>{errorMessage}</Text>
  }

  if (loading) {
    return <Loader center={true} />
  }

  return children
}
