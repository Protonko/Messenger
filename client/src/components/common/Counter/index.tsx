// types
import {Status} from 'models/common/status'

import {FC} from 'react'
import classNames from 'classnames'

export interface IPropsCounter {
  count: number | '99+',
  status?: Status,
}

export const Counter: FC<IPropsCounter> = ({
  count,
  status = Status.ACTIVE,
}) => {
  const classNameStatus = classNames(
    'counter',
    {'counter--muted': status === Status.MUTED}
  )

  return (
    <div className={classNameStatus}>
      <span className="counter__text">
        {count}
      </span>
    </div>
  )
}
