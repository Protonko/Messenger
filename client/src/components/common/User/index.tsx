import type {FC} from 'react'
import classNames from 'classnames'
import {TextSize, TextTypes} from 'models/common/text'
import {Avatar, IPropsAvatar} from 'components/common/Avatar'
import {Text} from 'components/common/Text'

export interface IPropsUser extends IPropsAvatar {
  name: string
  additionalClassName?: string
  description?: string
}

export const User: FC<IPropsUser> = ({
  name,
  src,
  description,
  size,
  additionalClassName,
}) => {
  const classNamesUser = classNames('user', {
    [additionalClassName ?? '']: !!additionalClassName,
  })

  return (
    <div className={classNamesUser}>
      <Avatar size={size} name={name} src={src} />
      <div className="user__data">
        <Text
          type={TextTypes.h4}
          customStyles="user__data-name"
          numberOfLines={1}
        >
          {name}
        </Text>

        <Text
          type={TextTypes.mixed}
          size={TextSize.EXTRA_SMALL}
          customStyles="user__data-description"
          numberOfLines={1}
        >
          {description ?? ''}
        </Text>
      </div>
    </div>
  )
}
