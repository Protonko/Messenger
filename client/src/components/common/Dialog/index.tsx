import type {FC} from 'react'
import type {IDialog} from 'models/dialog'
import {ReadStatus} from 'models/common/status'
import {TextSize, TextTypes} from 'models/common/text'
import {ReactComponent as DoubleCheck} from 'assets/icons/double-check.svg'
import {ReactComponent as Check} from 'assets/icons/check.svg'
import {COLORS} from 'static/colors'
import {Text} from 'components/common/Text'
import {Counter} from 'components/common/Counter'
import {Avatar} from 'components/common/Avatar'

export const Dialog: FC<IDialog> = ({
  name,
  description,
  avatar,
  edited,
  createdAt,
  updatedAt,
  messages,
  status,
  readStatus,
}) => {
  const counter = messages.toString().length > 2 ? '99+' : messages

  const renderStatus = () => {
    if (!!counter) {
      return <Counter count={counter} status={status}/>
    }

    if (readStatus) {
      const props = {
        width: 15,
        height: 15,
        color: COLORS.dustyGray,
      }
      return readStatus === ReadStatus.READ ? <DoubleCheck {...props} /> : <Check {...props} />
    }

    return null
  }

  return (
    <div className="dialog">
      <Avatar src={avatar} />

      <div className="dialog__data">
        <div className="dialog__text">
          <Text
            type={TextTypes.h4}
            customStyles="dialog__text-title"
            numberOfLines={1}
          >
            {name}
          </Text>
          <Text>{description}</Text>
        </div>

        <div className="dialog__info">
          <Text
            type={TextTypes.mixed}
            size={TextSize.EXTRA_SMALL}
            customStyles="dialog__info-date"
          >
            {new Date(edited ? updatedAt : createdAt).toLocaleDateString()}
          </Text>

          {renderStatus()}
        </div>
      </div>
    </div>
  )
}
