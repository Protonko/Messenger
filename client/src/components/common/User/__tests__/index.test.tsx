import {User, IPropsUser} from 'components/common/User'
import {mount, ReactWrapper} from 'enzyme'

const ELEMENT_SELECTORS = {
  user: 'user',
  description: 'user__data-description',
  custom: 'custom',
}

describe('User', () => {
  let component: ReactWrapper<IPropsUser>
  let props: IPropsUser

  beforeEach(() => {
    props = {
      name: 'Test Name',
      src: 'test/src',
    }
  })

  it('Should render User component', () => {
    component = mount(<User {...props} />)

    expect(component.find(`.${ELEMENT_SELECTORS.user}`).length).toBe(1)
    expect(component).toMatchSnapshot('User')
  })

  it('Should render User component with description', () => {
    props.description = 'test description'
    component = mount(<User {...props} />)

    expect(component.find(`.${ELEMENT_SELECTORS.description}`).text()).toBe(props.description)
    expect(component).toMatchSnapshot('User with description')
  })

  it('Should render User component with custom styles', () => {
    props.customStyles = ELEMENT_SELECTORS.custom
    component = mount(<User {...props} />)

    expect(component.find(`.${ELEMENT_SELECTORS.custom}`).length).toBe(1)
    expect(component).toMatchSnapshot('User with custom styles')
  })
})
