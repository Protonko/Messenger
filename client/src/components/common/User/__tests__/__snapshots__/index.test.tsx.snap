// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Should render User component with custom styles: User with custom styles 1`] = `
<User
  customStyles="custom"
  name="Test Name"
  src="test/src"
>
  <div
    className="user custom"
  >
    <Avatar
      name="Test Name"
      src="test/src"
    >
      <div
        className="avatar"
      >
        <img
          alt=""
          className="avatar__image"
          src="test/src"
        />
        <Loader>
          <div
            className="loader"
          >
            <div
              className="loader__line"
              key="0"
            />
            <div
              className="loader__line"
              key="1"
            />
            <div
              className="loader__line"
              key="2"
            />
            <div
              className="loader__line"
              key="3"
            />
            <div
              className="loader__line"
              key="4"
            />
            <div
              className="loader__line"
              key="5"
            />
            <div
              className="loader__line"
              key="6"
            />
            <div
              className="loader__line"
              key="7"
            />
            <div
              className="loader__line"
              key="8"
            />
            <div
              className="loader__line"
              key="9"
            />
            <div
              className="loader__line"
              key="10"
            />
            <div
              className="loader__line"
              key="11"
            />
          </div>
        </Loader>
      </div>
    </Avatar>
    <div
      className="user__data"
    >
      <Component
        customStyles="user__data-name"
        numberOfLines={1}
        type="h4"
      >
        <Dotdotdot
          clamp={1}
          tagName="div"
          truncationChar="…"
          useNativeClamp={false}
        >
          <div>
            <Component
              customStyles="user__data-name"
              type="h4"
            >
              <h4
                className="text text--heading-4 user__data-name"
              >
                Test Name
              </h4>
            </Component>
          </div>
        </Dotdotdot>
      </Component>
      <Component
        customStyles="user__data-description"
        numberOfLines={1}
        size="XS"
        type="mixed"
      >
        <Dotdotdot
          clamp={1}
          tagName="div"
          truncationChar="…"
          useNativeClamp={false}
        >
          <div>
            <Component
              customStyles="user__data-description"
              size="XS"
              type="mixed"
            >
              <p
                className="text text--xs user__data-description"
              />
            </Component>
          </div>
        </Dotdotdot>
      </Component>
    </div>
  </div>
</User>
`;

exports[`User Should render User component with description: User with description 1`] = `
<User
  description="test description"
  name="Test Name"
  src="test/src"
>
  <div
    className="user"
  >
    <Avatar
      name="Test Name"
      src="test/src"
    >
      <div
        className="avatar"
      >
        <img
          alt=""
          className="avatar__image"
          src="test/src"
        />
        <Loader>
          <div
            className="loader"
          >
            <div
              className="loader__line"
              key="0"
            />
            <div
              className="loader__line"
              key="1"
            />
            <div
              className="loader__line"
              key="2"
            />
            <div
              className="loader__line"
              key="3"
            />
            <div
              className="loader__line"
              key="4"
            />
            <div
              className="loader__line"
              key="5"
            />
            <div
              className="loader__line"
              key="6"
            />
            <div
              className="loader__line"
              key="7"
            />
            <div
              className="loader__line"
              key="8"
            />
            <div
              className="loader__line"
              key="9"
            />
            <div
              className="loader__line"
              key="10"
            />
            <div
              className="loader__line"
              key="11"
            />
          </div>
        </Loader>
      </div>
    </Avatar>
    <div
      className="user__data"
    >
      <Component
        customStyles="user__data-name"
        numberOfLines={1}
        type="h4"
      >
        <Dotdotdot
          clamp={1}
          tagName="div"
          truncationChar="…"
          useNativeClamp={false}
        >
          <div>
            <Component
              customStyles="user__data-name"
              type="h4"
            >
              <h4
                className="text text--heading-4 user__data-name"
              >
                Test Name
              </h4>
            </Component>
          </div>
        </Dotdotdot>
      </Component>
      <Component
        customStyles="user__data-description"
        numberOfLines={1}
        size="XS"
        type="mixed"
      >
        <Dotdotdot
          clamp={1}
          tagName="div"
          truncationChar="…"
          useNativeClamp={false}
        >
          <div>
            <Component
              customStyles="user__data-description"
              size="XS"
              type="mixed"
            >
              <p
                className="text text--xs user__data-description"
              >
                test description
              </p>
            </Component>
          </div>
        </Dotdotdot>
      </Component>
    </div>
  </div>
</User>
`;

exports[`User Should render User component: User 1`] = `
<User
  name="Test Name"
  src="test/src"
>
  <div
    className="user"
  >
    <Avatar
      name="Test Name"
      src="test/src"
    >
      <div
        className="avatar"
      >
        <img
          alt=""
          className="avatar__image"
          src="test/src"
        />
        <Loader>
          <div
            className="loader"
          >
            <div
              className="loader__line"
              key="0"
            />
            <div
              className="loader__line"
              key="1"
            />
            <div
              className="loader__line"
              key="2"
            />
            <div
              className="loader__line"
              key="3"
            />
            <div
              className="loader__line"
              key="4"
            />
            <div
              className="loader__line"
              key="5"
            />
            <div
              className="loader__line"
              key="6"
            />
            <div
              className="loader__line"
              key="7"
            />
            <div
              className="loader__line"
              key="8"
            />
            <div
              className="loader__line"
              key="9"
            />
            <div
              className="loader__line"
              key="10"
            />
            <div
              className="loader__line"
              key="11"
            />
          </div>
        </Loader>
      </div>
    </Avatar>
    <div
      className="user__data"
    >
      <Component
        customStyles="user__data-name"
        numberOfLines={1}
        type="h4"
      >
        <Dotdotdot
          clamp={1}
          tagName="div"
          truncationChar="…"
          useNativeClamp={false}
        >
          <div>
            <Component
              customStyles="user__data-name"
              type="h4"
            >
              <h4
                className="text text--heading-4 user__data-name"
              >
                Test Name
              </h4>
            </Component>
          </div>
        </Dotdotdot>
      </Component>
      <Component
        customStyles="user__data-description"
        numberOfLines={1}
        size="XS"
        type="mixed"
      >
        <Dotdotdot
          clamp={1}
          tagName="div"
          truncationChar="…"
          useNativeClamp={false}
        >
          <div>
            <Component
              customStyles="user__data-description"
              size="XS"
              type="mixed"
            >
              <p
                className="text text--xs user__data-description"
              />
            </Component>
          </div>
        </Dotdotdot>
      </Component>
    </div>
  </div>
</User>
`;
